File: wlr/SPX.Paths/StepLookup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using SPX.Core;

namespace SPX.Paths
{
    class StepLookup
    {
        private static Dictionary<Point2,Point2>[] s_lookups = new Dictionary<Point2,Point2>[1000];
        private static int m_index = 0;

        public static Dictionary<Point2,Point2> Copy(Dictionary<Point2,Point2> walk)
        {
            if (s_lookups[0] == null)
            {
                for (int ii = 0; ii < s_lookups.Count(); ii++)
                {
                    s_lookups[ii] = new Dictionary<Point2,Point2>(200);
                }
            }
            m_index = (m_index + 1) % s_lookups.Count();
            s_lookups[m_index].Clear();
            foreach (var keyval in walk)
            {
                s_lookups[m_index].Add(keyval.Key,keyval.Value);
            }
            return s_lookups[m_index];
        }
    }
}

File: code/spx-ported/src/com/spx/paths/StepLookup.java
package com.spx.paths;import com.xna.wrapper.*;import java.util.*;import com.spx.core.*;
    class StepLookup
    {
        private static ArrayList<HashMap<Point2,Point2>> __lookups = new ArrayList<HashMap<Point2,Point2>>();
        private static int _index = 0;        private static final int __lookupCount = 1000;

        public static HashMap<Point2,Point2> Copy(HashMap<Point2,Point2> walk)
        {
            if (__lookups.size() == 0)
            {
                for (int ii = 0; ii < __lookupCount; ii++)
                {
                    __lookups.add(new HashMap<Point2,Point2>(200));
                }
            }
            _index = (_index + 1) % __lookups.size();
            __lookups.get(_index).clear();
            for(Point2 key:walk.keySet())
            {
                __lookups.get(_index).put(key,walk.get(key));
            }
            return __lookups.get(_index);
        }
    }
